pkg/scheduler/objects/application.go

```diff
@@ -827,13 +827,41 @@ func (sa *Application) tryPlaceholderAllocate(nodeIterator func() NodeIterator,
 			if ph.released || request.taskGroupName != ph.taskGroupName {
 				continue
 			}
+			// before we check anything we need to check the resources equality
+			delta := resources.Sub(ph.AllocatedResource, request.AllocatedResource)
+			// Any negative value in the delta means that at least one of the requested resource in the real
+			// allocation is larger than the placeholder. We need to cancel the placeholder and check the next
+			// placeholder. This should trigger a cleanup of all the placeholder as a task group is always one size.
+			if delta.HasNegativeValue() {
+				log.Logger().Warn("releasing placeholder: real allocation is larger than placeholder",
+					zap.String("requested resource", request.AllocatedResource.String()),
+					zap.String("placeholderID", ph.UUID),
+					zap.String("placeholder resource", ph.AllocatedResource.String()))
+				// release the placeholder and tell the RM
+				ph.released = true
+				sa.notifyRMAllocationReleased(sa.rmID, []*Allocation{ph}, si.TerminationType_TIMEOUT, "cancel placeholder: resource incompatible")
```

Comment:
`si.TerminationType_TIMEOUT` - can this be changed to something else or would we break API with that? Something like `si.TerminationType_RESOURCE_MISMATCH`.

Reply:
I did not introduce a new state as the processing for both would be exactly the same in both the shim and the core. It would also require a new SI and a new shim to make the same changes work.